src/middlewared/middlewared/plugins/truenas_connect/finalize_registration.py

```diff
@@ -0,0 +1,116 @@
+                # TODO: Add alerts
+                logger.debug('TNC claim token has expired')
+                await self.status_update(Status.REGISTRATION_FINALIZATION_TIMEOUT)
+                return
+
+            try:
+                logger.debug('Attempt %r: Polling for TNC registration finalization', try_num)
+                status = await self.poll_once(claim_token, system_id)
+            except asyncio.CancelledError:
+                await self.status_update(
+                    Status.REGISTRATION_FINALIZATION_TIMEOUT, 'TNC registration finalization polling has been cancelled'
+                )
+                raise
+            except Exception as e:
+                logger.debug('TNC registration has not been finalized yet: %r', str(e))
+                status = {'error': str(e)}
+            finally:
+                try_num += 1
+
+            if status['error'] is None:
+                # We have got the key now and the registration has been finalized
+                if 'token' not in status['response']:
+                    logger.error(
+                        'Registration finalization failed for TNC as token not found in response: %r',
+                        status['response']
+                    )
+                    await self.status_update(Status.REGISTRATION_FINALIZATION_FAILED)
+                else:
+                    token = status['response']['token']
+                    decoded_token = {}
+                    try:
+                        decoded_token = jwt.decode(token, options={'verify_signature': False})
+                    except jwt.exceptions.DecodeError:
+                        logger.error('Invalid JWT token received from TNC')
+                        await self.status_update(Status.REGISTRATION_FINALIZATION_FAILED)
+                        return
+                    else:
+                        if diff := {'account_id', 'system_id'} - set(decoded_token):
+                            logger.error('JWT token does not contain required fields: %r', diff)
+                            await self.status_update(Status.REGISTRATION_FINALIZATION_FAILED)
+                            return
+
+                    await self.middleware.call(
+                        'datastore.update', 'truenas_connect', config['id'], {
+                            'jwt_token': token,
+                            'registration_details': decoded_token,
+                        }
+                    )
+                    await self.status_update(Status.CERT_GENERATION_IN_PROGRESS)
+                    logger.debug('TNC registration has been finalized')
+                    self.middleware.create_task(self.middleware.call('tn_connect.acme.initiate_cert_generation'))
+                    # Remove claim token from cache
+                    await self.middleware.call('cache.pop', CLAIM_TOKEN_CACHE_KEY)
+                    return
+
+            await asyncio.sleep(self.POLLING_GAP_MINUTES * 60)
```

Comment:
Can this be a little bit more responsive? Like every 1min?

Assuming this is called when the user opens the register page, it will always fail first, then in most cases (90-95%), it will succeed after a 1-2min delay after login and registration, when this is called again. More often then not there will be a needless delay.